pkg/controller/endpointslice/endpointslice_controller.go

```diff
@@ -302,20 +307,35 @@ func (c *Controller) syncService(key string) error {
 	return nil
 }
 
-// obj could be a *v1.Service or a DeletionalFinalStateUnknown marker item
+// obj could be a *v1.Service.
 func (c *Controller) enqueueService(obj interface{}) {
+	if key, err := enqueueKey(obj, c.queue); err == nil {
+		_ = c.serviceSelectorCache.Update(key, obj.(*v1.Service))
+	}
+}
+
+// obj could be a DeletionalFinalStateUnknown marker item.
+func (c *Controller) dequeueService(obj interface{}) {
+	if key, err := enqueueKey(obj, c.queue); err == nil {
+		c.serviceSelectorCache.Delete(key)
+	}
+}
+
+// obj could be an *v1.Service, or a DeletionFinalStateUnknown marker item.
+func enqueueKey(obj interface{}, queue workqueue.RateLimitingInterface) (string, error) {
```

Comment:
With the functions renamed and the logic inside reordered, I think this can go back to be the original simpler `enqueueService` method.